### PR TITLE
python docs: add an example for a virtualenv and pip through nix-shell

### PR DESCRIPTION
**Pull request by Azulinho**
###### Motivation for this change

To address:
https://github.com/NixOS/nixpkgs/issues/66

And to document what actually worked for me while dealing with different python projects that required a virtualenv and compilation of external dependencies.

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: doc/languages-frameworks/python.md [hunk 1]
```diff
@@ -804,6 +804,43 @@ If you want to create a Python environment for development, then the recommended
...
+### How to consume python modules using pip in a virtualenv like I am used to on other Operating Systems ?
+
+This is an example of a `default.nix` for a `nix-shell`, which allows to consume a `virtualenv` environment,
+and install python modules through `pip` the traditional way.
+
+Create this `default.nix` file, together with a `requirements.txt` and simply execute `nix-shell`.
+
+```
+
+th import <nixpkgs> {};
```
Comment (FRidh) [inline doc/languages-frameworks/python.md]: 2 letters missing here

File: doc/languages-frameworks/python.md [hunk 2]
```diff
@@ -804,6 +804,43 @@ If you want to create a Python environment for development, then the recommended
...
+with pkgs.python27Packages;
+
+buildPythonPackage {
```
Comment (FRidh) [inline doc/languages-frameworks/python.md]: we're not building a Python package, but an environment. I suggest you use `python.buildEnv` or else I think you might as well use `stdenv.mkDerivation`.

File: doc/languages-frameworks/python.md [hunk 3]
```diff
@@ -804,6 +804,43 @@ If you want to create a Python environment for development, then the recommended
...
+    openssl
+    git
+    libxml2
+    libxslt
+    libzip
+    python27Full
+    python27Packages.virtualenv
+    python27Packages.pip
+    stdenv
+    zlib ];
```
Comment (FRidh) [inline doc/languages-frameworks/python.md]: what's the purpose of all these packages? Are some hard dependencies, or are they just examples of packages?

File: doc/languages-frameworks/python.md [hunk 4]
```diff
@@ -804,6 +804,43 @@ If you want to create a Python environment for development, then the recommended
...
+  src = null;
+  # When used as `nix-shell --pure`
+  shellHook = ''
+  unset http_proxy
```
Comment (FRidh) [inline doc/languages-frameworks/python.md]: indentation of this block

**Review by FRidh — CHANGES_REQUESTED** [review]
How to create a functioning `virtualenv` on Nix is a recurring question and it's good to address this in the manual. 

When including this example it should also describe the limitations of this method. E.g., how are you dealing with packages that have extension types, like `numpy`?

How do you make sure the paths in the store remain available? Will you add a gc root? And what happens when you `nix-shell` again but `nixpkgs` has updated?

Can you use additional packages from Nix? And if so, how do you deal with dependencies between packages in the `virtualenv` and packages in the store? E.g., you make package `a` available via Nix, and install `b` via `requirements.txt`.

File: doc/languages-frameworks/python.md [hunk 5]
```diff
@@ -804,6 +804,43 @@ If you want to create a Python environment for development, then the recommended
...
+  export GIT_SSL_CAINFO=/etc/ssl/certs/ca-bundle.crt
```
Comment (FRidh) [inline doc/languages-frameworks/python.md]: Is this needed, and if so, why?
Maybe also `SOURCE_DATE_EPOCH` for when you want to build wheels?

File: doc/languages-frameworks/python.md (hunk 1 above)
Comment (Azulinho) [inline doc/languages-frameworks/python.md]: fixing it

File: doc/languages-frameworks/python.md (hunk 2 above)
Comment (Azulinho) [inline doc/languages-frameworks/python.md]: works for me

File: doc/languages-frameworks/python.md (hunk 3 above)
Comment (Azulinho) [inline doc/languages-frameworks/python.md]: my particular requirements.txt is for a twisted project which requires a few python modules containing C extensions. The packages listed are the dependencies required to build those particular python modules.

I guess I can expand that block stating which are required (like the virtualenv and pip), and which are related to dependencies of the modules listed in the requirements.txt file.

File: doc/languages-frameworks/python.md (hunk 5 above)
Comment (Azulinho) [inline doc/languages-frameworks/python.md]: my bad, those are actually only relevant to my own setup. pulling them out...
and adding SOURCE_DATE_EPOCH

**Comment by Mic92** [comment]
Thanks.

**Comment by Mic92** [comment]
I actually wanted to squash them together, but hit the wrong button...

**Comment by CMCDragonkai** [comment]
I'm always been confused here. The normal python docs mentions things like `withPackages`, `pkgs.python35Packages.python.buildEnv.override` (which are the same thing) and the only thing they do is create a python with the necessary nixpkgs environment packages. However you don't actually need this, you can just use a normal `shell.nix` with all the python dependencies placed into the `buildInputs` of `stdenv.mkDerivation` and all of the python packages will be made available to the nix-shell installed python using packageHooks that change the `PYTHONPATH` variable.

So then the question becomes how does pip interact within a nix-shell. By default it simply doesn't work. I would prefer that pip worked similar to Node's npm or PHP's composer or even Haskell's stack, in that dependencies were installed in the current directory. Unfortunately this is not possible without some hacks into `--root` or `--prefix` or `--install-option` which is quite complex as well.

The only way I've found that actually works is nix-shell for system dependencies and python dependencies that I don't want to compile via pip + virtualenv and pip for all the other normal pip dependencies. But I thought that there would be a shellHook or something that would change pip to install into the local directory and thus we wouldn't need an extra virtualenv layer in between nix-shell and pip.

There's also the question of what happens when there's a nix installed python dependency like numpy and a numpy specified in the `requirements.txt`. When pip runs an installation, does it respect the fact that numpy already exists, or does it check the version requirements and decides to install a local copy, and if it does, does the resulting python find the virtualenv + pip installed one or the nix supplied one?

File: doc/languages-frameworks/python.md [hunk 6]
```diff
@@ -804,6 +804,55 @@ If you want to create a Python environment for development, then the recommended
...
+  buildInputs = [
+    # these packages are required for virtualenv and pip to work:
+    #
+    python27Full
+    python27Packages.virtualenv
+    python27Packages.pip
+    # the following packages are related to the dependencies of your python 
+    # project. 
+    # In this particular example the python modules listed in the 
+    # requirements.tx require the following packages to be installed locally
```
Comment (Mic92) [inline doc/languages-frameworks/python.md]:
```diff
- requirements.tx
+ requirements.txt
```

**Pull request merged**